### PR TITLE
feat(parsers.influx): Allow a user to set the timestamp precision

### PR DESCRIPTION
**Pull request by powersj**
fixes: #13600

### CODE REVIEW

**Comment by Hipska** [comment]
This is nice to have, but it does not fix the specific request in #13407

**Comment by eth5201314** [comment]
I noticed that your license is unclear. Can you provide more information about how your code can be used?

File: plugins/inputs/influxdb_v2_listener/influxdb_v2_listener.go [hunk 1]
```diff
@@ -271,7 +271,10 @@ func (h *InfluxDBV2Listener) handleWrite() http.HandlerFunc {
 
 			if precisionStr != "" {
 				precision := getPrecisionMultiplier(precisionStr)
-				parser.SetTimePrecision(precision)
+				if err = parser.SetTimePrecision(precision); err != nil {
+					h.Log.Debugf("Error setting precision of parser: %v", err.Error())
```
Comment (srebhan) [inline plugins/inputs/influxdb_v2_listener/influxdb_v2_listener.go:275]:
```suggestion
					h.Log.Debugf("Error setting precision of parser: %v", err)
```

**Review by srebhan — COMMENTED** [review]
Just one small comment...

**Comment by srebhan** [comment]
@eth5201314 I'm not exactly sure what you are asking... Is your question related to Telegraf as a whole?

File: plugins/parsers/influx/influx_upstream/parser_test.go [hunk 2]
```diff
@@ -765,6 +766,115 @@ func TestSeriesParser(t *testing.T) {
...
+		})
+	}
+}
+
+func TestParserInvalidTimestampPrecision(t *testing.T) {
+	d := config.Duration(0)
+	for _, precision := range []string{"1h", "1d", "2s", "1m", "2ns"} {
+		require.NoError(t, d.UnmarshalText([]byte(precision)))
+		parser := Parser{InfluxTimestampPrecsion: d}
+		require.Error(t, parser.Init())
```
Comment (Hipska) [inline plugins/parsers/influx/influx_upstream/parser_test.go:874]:
```suggestion
		require.ErrorContains(t, parser.Init(), "invalid time precision")
```

File: plugins/parsers/influx/parser_test.go [hunk 3]
```diff
@@ -601,6 +603,116 @@ func TestParser(t *testing.T) {
...
+		})
+	}
+}
+
+func TestParserInvalidTimestampPrecision(t *testing.T) {
+	d := config.Duration(0)
+	for _, precision := range []string{"1h", "1d", "2s", "1m", "2ns"} {
+		require.NoError(t, d.UnmarshalText([]byte(precision)))
+		parser := Parser{InfluxTimestampPrecsion: d}
+		require.Error(t, parser.Init())
```
Comment (Hipska) [inline plugins/parsers/influx/parser_test.go:712]:
```suggestion
		require.ErrorContains(t, parser.Init(), "invalid time precision")
```

**Review by Hipska — CHANGES_REQUESTED** [review]

File: plugins/parsers/influx/parser_test.go [hunk 4]
```diff
@@ -3,6 +3,7 @@ package influx
 import (
 	"bytes"
 	"errors"
+	"fmt"
```
Comment (powersj) [inline plugins/parsers/influx/parser_test.go:6]:
```suggestion
```

File: plugins/parsers/influx/parser.go [hunk 5]
```diff
@@ -155,6 +157,15 @@ func (p *Parser) Init() error {
...
 
+	timeDuration := time.Duration(p.InfluxTimestampPrecsion)
+	switch timeDuration {
+	case 0:
+	case time.Nanosecond, time.Microsecond, time.Millisecond, time.Second:
+		p.SetTimePrecision(timeDuration)
+	default:
+		return fmt.Errorf("invalid time precision: %d", p.InfluxTimestampPrecsion)
+	}
+
```
[the remark this answers is not in the document]
Comment (srebhan) [reply plugins/parsers/influx/parser.go:168]: I think the issue is that the original influx parser is code-generated and therefore we cannot simply add/modify the function. I'm fine with the second instance...

[the remark this answers is not in the document]
Comment (Hipska) [reply plugins/parsers/influx/parser.go:168]: But why not align them to look the same?

**Review by Hipska — APPROVED** [review]

**Comment by Hipska** [comment]
Let's check after a new nightly build? I don't like that warning of file size change 😛

**Pull request merged**